README.md
## FPM-PLUGIN-TIANYI
用于接中国电信的NBIOT平台的插件

### Install
```bash
npm add fpm-plugin-tianyi --save
```

### Useage

- config
```
{
    "tianyi": {
        "appId": "****",
        "secret": "**",
        "webhook": "http://${****}/webhook/tianyi/notify/datachange",    // http的协议只能用于调试，正式环境还是需要切换成https
        "baseUrl": "https://device.api.ct10649.com:8743"    // This is the final tianyi iot gateway
    }
}
```

- subscribe

`#tianyi/notify`

- other

相关文档

[开发指南](https://180.101.147.208:8093/assets/docCenter/helpcenter/helpPortal/Portal/helpcenter.html?manualName=UserGuide_CMCC&docSite=CMCC&page=gettingStarted&lang=zh)

CHANGELOG.md
## 1.0.0 (2019-04-22)

Change the Tianyi Endpoint Api Address: `device.api.ct10649.com` replace the test address `180.101.147.89`.

## 0.0.1

Change `script` in `package.json`
`prepublishOnly` replace `prepublish` 

test/function.test.js
const { init, Func } = require("fpmc-jssdk");
const assert = require('assert');
init({ appkey:'123123', masterKey:'123123', endpoint: 'http://localhost:9999/api' });

describe('Function', function(){
  beforeEach(done => {
    done()
  })


  afterEach(done => {
    done()
  })

  it('Function A', async () => {
    try {
      const data = await new Func('tianyi.subscriptions').invoke({});
      console.log(data);
      assert.strictEqual(data == undefined, false, 'should not be undefined');
    } catch (error) {
      throw error;
    }
  });

  it('Function send', async () => {
    try {
      const data = await new Func('tianyi.send').invoke({
        deviceId: '141c0c85-24ca-4d2d-bf57-7e45591a3b88',
        command: { "serviceId": "Cmd", "method": "Set_Cmd", "paras": {"EXTRA":2,"LENGTH":8,"FN":3,"DATA_1": -66052,"DATA_2":4} } 
      });
      console.log(data);
      assert.strictEqual(data == undefined, false, 'should not be undefined');
    } catch (error) {
      throw error;
    }
  });
})


kit/cmd.js
const _ = require('lodash');
const assert = require('assert');
const { postJson } = require('./httpsutil');
const debug = require('debug')('fpm-plugin-tianyi:cmd');

const { cmdBaseUrl } = require('./setting');

const sendCmd = async ( args ) => {
  const { appId, accessToken, deviceId, command } = args;
  try {
    const url = `${ cmdBaseUrl }deviceCommands`;
    const rsp = await postJson({
      url,
      header: {
        app_key: appId,
        Authorization: `Bearer ${ accessToken }`
      },
      data: { deviceId, command }
    });
    const { code, commandId } = rsp;
    assert( code == 201, JSON.stringify(rsp));
    assert( commandId != undefined, JSON.stringify(rsp));
    return rsp;
  } catch (error) {
    debug('%O', error);
    throw error;
  }
};

exports.sendCmd = sendCmd;

src/biz.js
const _ = require('lodash');
const assert = require('assert');
const debug = require('debug')('fpm-plugin-tianyi:biz');
const { getToken } = require('../kit/auth');
const { sendCmd } = require('../kit/cmd');
const { defineUrlPrefix } = require('../kit/httpsutil');
const { subscriptions, subscribe } = require('../kit/subscribe');

module.exports = (fpm, options) => {

  const { appId, secret, baseUrl } = options;
  defineUrlPrefix(baseUrl);
  // handle the notify webhook
  fpm.subscribe(`#webhook/tianyi/notify`, (topic, message) => {
    // console.log(topic, message);
    debug('%o, %o', topic, message);
    fpm.publish('#tianyi/notify', message);
  })
  
  return {
    registerDevice: async ( args ) => {
      // register a device
      debug('%O', args)
      throw Error('stup')
    },

    send: async ( args ) => {
      // send data to device from server
      debug('%O', args)
      const { deviceId , command } = args;
      try {
        const { accessToken } = await getToken( { appId, secret } );
        const data = await sendCmd({
          appId,
          accessToken,
          deviceId: deviceId || '1d37de7c-8c01-4700-b6d3-0da16ece7a4b',
          command: command || { serviceId: `Light`, method: `SET_LEVEL`, paras: { value: 02 } }
        });
        const { commandId, code } = data;
        assert( code == 201, 'Code should be 201');
        assert( commandId != undefined, 'commandId should not be empty');
        return data;
      } catch (error) {
        throw error;
      }
    },
    subscribe: async ( args ) => {
      // subscribe a event
      debug('%O', args)
      const { notifyType = 'deviceDataChanged', callbackUrl } = args;
      try {
        const { accessToken } = await getToken( { appId, secret } );
        const data = await subscribe({
          appId,
          accessToken,
          notifyType,
          callbackUrl,
          
        });
        return data;
      } catch (error) {
        debug('error %O', error)
        return Promise.reject({
          message: error.message,
        })
      }
    },
    subscriptions: async ( args ) => {
      // get all of the subscriptions
      debug('%O', args)
      const { notifyType = 'deviceDataChanged', pageNo = 0, pageSize = 100 } = args;
      try {
        const { accessToken } = await getToken( { appId, secret } );
        const data = await subscriptions({
          appId,
          accessToken,
          notifyType,
          pageNo,
          pageSize,
        });
        return data;
      } catch (error) {
        debug('error %O', error)
        return Promise.reject({
          message: error.message,
        })
      }
    },

  }

}

kit/auth.js
const _ = require('lodash');
const assert = require('assert');
const fs = require('fs');
const path = require('path');
const debug = require('debug')('fpm-plugin-tianyi:auth');
const { postJson, postForm } = require('./httpsutil');

const { secBaseUrl } = require('./setting');

const LOGIN_URL = `${ secBaseUrl }login`;
const REFRESH_URL = `${ secBaseUrl }refreshToken`;

const CWD = process.cwd();
const tokenFile = path.join(CWD, 'token.json');

const getTokenInfo = () => {
  if(fs.existsSync(tokenFile)){
    const content = fs.readFileSync(tokenFile);
    try {
      return JSON.parse(content.toString());
    } catch (error) {
      debug('getTokenInfo() error! %o', error);
      return
    }
  }
  return;
}

const saveTokenInfo = token => {
  const ws = fs.createWriteStream(tokenFile);
  return new Promise( (resolve, reject) => {
    ws.write(JSON.stringify(token));
    ws.end();

    ws.on('error', (err) => { reject(err); });
    ws.on('finish', () => { resolve(1); });
})
}

let CACHE_TOKEN = getTokenInfo();

const checkToken = ( token ) => {
  try {
    const { code } = token;
    assert(code == 200, `${JSON.stringify(token)}`);    
    return true;
  } catch (error) {
    debug('Token is error! %o', error);
    return false;
  }
}

const getToken = async ({ appId, secret }) => {
  try {
    let token;
    const NOW = _.now();
    if(!_.isEmpty(CACHE_TOKEN)){
      const { expiresTime } = CACHE_TOKEN;
      if(!checkToken(token)){
        // 已存在的token 是错误的，无法使用，需要重新登录
        token = await postForm({ url: LOGIN_URL, data: { appId, secret }});
      }else{
        // token 正确
        if(NOW >= expiresTime){
          // invalid , refresh the token, save it, return it
          token = await refreshToken(CACHE_TOKEN);
          if(!checkToken(token)){
            // 已存在的 refreshCode 完全实效了，无法使用，需要重新登录
            token = await postForm({ url: LOGIN_URL, data: { appId, secret }});
          }
        }else{
          return CACHE_TOKEN;
        }
      }
    }else{
      token = await postForm({ url: LOGIN_URL, data: { appId, secret }});
    }
    assert( token != undefined, JSON.stringify({ code: -999, message: 'Get/Refresh Token Error!' }) );
    assert(checkToken(token) === true, JSON.stringify(token))
    const { expiresIn } = token;
    token.expiresTime = NOW + (expiresIn * 1000);
    token.appId = appId;
    token.secret = secret;
    // stash the token cached
    CACHE_TOKEN = token;
    await saveTokenInfo(token);
    return token;
  } catch (error) {
    debug('getToken() %o', error)
    throw error;
  }
};

const refreshToken = async ( arg ) => {
  try {
    const { refreshToken, appId, secret } = arg;
    assert( refreshToken != undefined, JSON.stringify({ code: -999, message: 'Refresh Token Required when invoke the refreshToken()' }))
    const token = await postJson({ url: REFRESH_URL, data: { appId, secret, refreshToken }});
    return token;
  } catch (error) {
    debug('refreshToken() %o', error)
    throw error;
  }
};

exports.getToken = getToken;
exports.refreshToken = refreshToken;

src/index.js
const _ = require('lodash');
const pkg = require('../package.json');
const Biz = require('./biz.js');

const debug = require('debug')(pkg.name);

module.exports = {
  bind: (fpm) => {
    const options = _.assign({
        baseUrl: 'https://device.api.ct10649.com:8743'
      },
      fpm.getConfig('tianyi', {
        appId: 'XdbNkfbFSRYPJhs4H2EJ4G12w6Aa',
        secret: 'N5fUGGNl9aJIDapXNVkI3SVIfCYa',
        webhook: 'https://iot.yunplus.io/webhook/tianyi/notify/datachange',
        baseUrl: 'https://device.api.ct10649.com:8743',
      }) 
    )

    const bizModule = Biz(fpm, options);

    const subscribe = async () => {
      try {
        const data = await bizModule.subscribe({ notifyType: 'deviceDataChanged', callbackUrl: options.webhook });
        return data;
      } catch (error) {
        debug('error %O', error)
        return Promise.reject({ message: error.message, });
      }
    }
    fpm.registerAction('BEFORE_SERVER_START', () => {
      fpm.extendModule('tianyi', bizModule);

      // make sure the webhook url has been subscribed!
      bizModule.subscriptions({notifyType: 'deviceDataChanged'})
        .then(data => {
          const { totalCount, subscriptions } = data;
          if(totalCount < 1){
            subscribe();
            return;
          }
          if(undefined === _.find(subscriptions, { callbackUrl: options.webhook })){
            subscribe();
            return;
          }
        })
        .catch(err => {
          debug('ERROR: %O', err);
        })
      //
    })
    return bizModule;
  }
}


kit/subscribe.js
const _ = require('lodash');
const assert = require('assert');
const { getJson, postJson } = require('./httpsutil');
const debug = require('debug')('fpm-plugin-tianyi:subscribe');

const { subBaseUrl } = require('./setting');

const subscribe = async ( args ) => {
  // subscribe a callback
  debug('%O', args)
  const { appId, accessToken, callbackUrl, notifyType } = args;
  try {
    assert(!_.isEmpty(callbackUrl), JSON.stringify({ code: -999, message: 'CallbackUrl Required!'}))
    assert(!_.isEmpty(notifyType), JSON.stringify({ code: -999, message: 'notifyType Required!'}))
    const url = `${ subBaseUrl }subscriptions`;
    const rsp = await postJson({
      url,
      header: {
        app_key: appId,
        Authorization: `Bearer ${ accessToken }`
      },
      data: { callbackUrl, notifyType }
    });
    const { code } = rsp;
    assert( code == 201, JSON.stringify(rsp));
    return rsp;
  } catch (error) {
    debug('%O', error);
    throw error;
  }
};

const subscriptions = async ( args ) => {
  // get all of the subscriptions
  debug('%O', args)
  const { appId, accessToken, notifyType = 'deviceDataChanged', pageNo = 0, pageSize = 100 } = args;
  try {
    const rsp = await getJson({
      url: `${subBaseUrl}subscriptions`,
      header: {
        app_key: appId,
        Authorization: `Bearer ${ accessToken }`
      },
      data: {
        notifyType,
        pageNo,
        pageSize,
      }
    })
    const { code } = rsp;
    assert( code == 200, JSON.stringify(rsp));
    return rsp;
  } catch (error) {
    throw error;
  }
};

exports.subscribe = subscribe;
exports.subscriptions = subscriptions;

kit/httpsutil.js
const _ = require('lodash');
const assert = require('assert');
const debug = require('debug')('fpm-plugin-tianyi:httpsutil');
const fs = require('fs');
const request = require('request');
const path = require('path');
const querystring = require('querystring');

const ContentType_Form = {
  'content-type': 'application/x-www-form-urlencoded',
};

const ContentType_Json = {
  'content-type': 'application/json',
};

let urlPrefix = 'https://device.api.ct10649.com:8743'

const defineUrlPrefix = ( prefix ) => {
  urlPrefix = prefix;
  debug('new urlprefix defined %s', urlPrefix);
}
const _options = {
  url: null,
  method: 'POST',
  agentOptions: {
    cert: fs.readFileSync(path.join(__dirname, '../cert/client.crt')),
    key: fs.readFileSync(path.join(__dirname, '../cert/client.key')),
  },
  json: true,
  headers: null,
  strictSSL: false
};

const doRequest = ( options ) => {
  debug('%O', options)

  return new Promise(( rs, rj) => {
    try {
      assert(!!options.url, 'URL required!') 
    } catch (error) {
      rj({ error })
      return;
    }
    if(!_.startsWith(options.url, 'http')){
      options.url = urlPrefix + options.url;
    }
    request(options, (error, response, body) => {
      if(error){
        rj({ error })
        return;
      }
      const { statusCode } = response;
      try {
        rs(_.assign(body, { code: statusCode }) );
      } catch (error) {
        rs({ body, code: statusCode });
      }
    });
  });
}

const postCreator = (type = 'Form' ) => {
  const defaultHeader = type === 'Form' ? ContentType_Form: ContentType_Json;
  return args => {
    const { url, data, header } = args;
    assert( !_.isEmpty(url), JSON.stringify({ code: -999, message: `URL: ${ url } required ~`}));
    const options = _.assign({}, _options, { url, headers: _.assign(defaultHeader , header) })
    if( type === 'Form'){
      options.form = data;
    }else{
      options.body = data;
    }
    return doRequest(options);
  }
}

/**
 * create a GET requrest.
 */
const getJson = (args) => {
  const { url, data, header } = args;
  let params;
  if(!_.isEmpty(data)){
    params = querystring.stringify(data);
  }
  assert( !_.isEmpty(url), JSON.stringify({ code: -999, message: `URL: ${ url } required ~`}));
  const options = _.assign({}, _options, { url: `${url}?${ params || '' }`, method: 'GET', headers: _.assign(ContentType_Json , header) })
  return doRequest(options);
}
exports.postJson = postCreator( 'Json' );
exports.postForm = postCreator( 'Form' );
exports.defineUrlPrefix = defineUrlPrefix;
exports.getJson = getJson;